runtime/src/iree-amd-aie/driver/xrt/native_executable.cc

```diff
@@ -0,0 +1,257 @@
+  iree_hal_xrt_kernel_params_t entry_points[];
+} iree_hal_xrt_native_executable_t;
+
+namespace {
+extern const iree_hal_executable_vtable_t iree_hal_xrt_native_executable_vtable;
+}  // namespace
+
+static iree_hal_xrt_native_executable_t* iree_hal_xrt_native_executable_cast(
+    iree_hal_executable_t* base_value) {
+  IREE_HAL_ASSERT_TYPE(base_value, &iree_hal_xrt_native_executable_vtable);
+  return (iree_hal_xrt_native_executable_t*)base_value;
+}
+
+// Verifies the structure of the flatbuffer so that we can avoid doing so during
+// runtime.
+//
+// There are still some conditions we must be aware of (such as omitted names on
+// functions with internal linkage), however we shouldn't need to bounds check
+// anything within the flatbuffer after this succeeds.
+static iree_status_t iree_amd_aie_hal_xrt_native_executable_flatbuffer_verify(
+    iree_const_byte_span_t flatbuffer_data) {
+  if (!flatbuffer_data.data || flatbuffer_data.data_length < 16) {
+    return iree_make_status(
+        IREE_STATUS_INVALID_ARGUMENT,
+        "flatbuffer data is not present or less than 16 bytes (%zu total)",
+        flatbuffer_data.data_length);
+  }
+
+  // Run flatcc generated verification. This ensures all pointers are in-bounds
+  // and that we can safely walk the file, but not that the actual contents of
+  // the flatbuffer meet our expectations.
+  int verify_ret = iree_amd_aie_hal_xrt_ExecutableDef_verify_as_root(
+      flatbuffer_data.data, flatbuffer_data.data_length);
+  if (verify_ret != flatcc_verify_ok) {
+    return iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                            "flatbuffer verification failed: %s",
+                            flatcc_verify_error_string(verify_ret));
+  }
+
+  iree_amd_aie_hal_xrt_ExecutableDef_table_t executable_def =
+      iree_amd_aie_hal_xrt_ExecutableDef_as_root(flatbuffer_data.data);
+
+  flatbuffers_string_vec_t entry_points_vec =
+      iree_amd_aie_hal_xrt_ExecutableDef_entry_points_get(executable_def);
+  size_t entry_point_count = flatbuffers_string_vec_len(entry_points_vec);
+  for (size_t i = 0; i < entry_point_count; ++i) {
+    if (!flatbuffers_string_len(
+            flatbuffers_string_vec_at(entry_points_vec, i))) {
+      return iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                              "executable entry point %zu has no name", i);
+    }
+  }
+
+  flatbuffers_string_t xclbin =
+      iree_amd_aie_hal_xrt_ExecutableDef_xclbin_get(executable_def);
+  if (flatbuffers_string_len(xclbin) == 0) {
+    return iree_make_status(IREE_STATUS_INVALID_ARGUMENT, "no xclbin present");
+  }
+
+  iree_amd_aie_hal_xrt_AsmInstDef_vec_t asm_instr =
+      iree_amd_aie_hal_xrt_ExecutableDef_asm_instrs_get(executable_def);
+  if (iree_amd_aie_hal_xrt_AsmInstDef_vec_len(asm_instr) != entry_point_count) {
+    return iree_make_status(
+        IREE_STATUS_INVALID_ARGUMENT,
+        "number of entry points and asm instructions mismatched");
```

Comment:
Can we also print out the number of entry points the number asm instructions to be clear?